### PR TITLE
Update ui tests to allow dockerized selenium

### PR DESCRIPTION
**Pull request by ericenns**
## Description of changes
UI testing has been refactored to enable use of a selenium/standalone-chrome docker container.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [ ] Tests added (or description of how to test) for any new features.
* [x] User documentation updated for UI or technical changes.

### CODE REVIEW

**Comment by ericenns** [comment]
Work in progress as uploads currently fail as the selenium container does not have access to the `/tmp` on the host machine, this could be fixed with changing how the container is started but it might be better to refactor where these tmp files are created.

**Comment by ericenns** [comment]
No longer work in progress, as file uploads now work with updated docs.

**Comment by ericenns** [comment]
@tom114 this is ready for you to review whenever you have a chance

File: CHANGELOG.md [hunk 1]
```diff
@@ -37,6 +37,7 @@ Changes
 * [UI/Developer]: Fixed issue with deleting a user group if it is linked to any projects.
 * [UI]: Fixes issue where importing an excel file that contained "Created Date" and "Modified Date" created metadata fields with those labels.
 * [REST]: Updated http status code returned (400 Bad Request) when uploading files to the wrong type of run
+* [Developer]: Updated UI testing to allow using an external selenium webdriver.
```
Comment (tom114) [inline CHANGELOG.md]: Merge `development` and move this to the new release section please.

File: doc/developer/interface/testing/index.md [hunk 2]
```diff
@@ -55,6 +55,20 @@ You need to match up the version of [Chrome] with [chromedriver]. For example, [
...
+### Using [selenium/standalone-chrome]
+
+To use [selenium/standalone-chrome] you will need to have docker installed on your machine.
+
+Note: the following command assumes you are in the root directory of the repository
+
+First start up the [selenium/standalone-chrome] container:
+```bash
+mkdir -p /tmp/irida
+docker run -d -p 4444:4444 -v $PWD:$PWD -v /tmp/irida:/tmp/irida -v /dev/shm:/dev/shm selenium/standalone-chrome:3.6
```
[the remark this answers is not in the document]
Comment (ericenns) [reply doc/developer/interface/testing/index.md]: There is no reason that it needs 3.6 specifically, I was just using that initially as it helped me to debug the error that we were getting with ui tests on travis ci. I will update this to `latest` tag

File: src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/remoteapi/CreateRemoteAPIPageIT.java [hunk 3]
```diff
@@ -57,8 +57,8 @@ public void testCreateClientWithSpacesInFront() {
 
 	@Test
 	public void testAndConnectToClient() {
-		String applicationPort = page.getApplicationPort();
-		String url = "http://localhost:" + applicationPort + "/api";
+		String baseUrl = page.getBaseUrl();
+		String url = baseUrl + "api";
```
[the remark this answers is not in the document]
Comment (tom114) [reply src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/remoteapi/CreateRemoteAPIPageIT.java]: After discussion we're going to skip this.  We build URLs like this all over the tests already.  It's just in test so no big deal.

File: run-tests.sh [hunk 4]
```diff
@@ -109,8 +113,17 @@ test_rest() {
 }
 
 test_ui() {
-	mvn clean verify -B -Pui_testing -Dwebdriver.chrome.driver=$CHROME_DRIVER -Dirida.it.nosandbox=true -Dirida.it.headless=$HEADLESS -Djdbc.url=$JDBC_URL -Dirida.it.rootdirectory=$TMP_DIRECTORY -Dsequence.file.base.directory=$SEQUENCE_FILE_DIR -Dreference.file.base.directory=$REFERENCE_FILE_DIR -Doutput.file.base.directory=$OUTPUT_FILE_DIR -Djdbc.pool.maxWait=$DB_MAX_WAIT_MILLIS $@
+    SELENIUM_OPTS=""
+    if [ "$SELENIUM_DOCKER" = false ];
+    then
+        SELENIUM_OPTS="-Dwebdriver.chrome.driver=$CHROME_DRIVER"
+    else
+        docker run -d -p 4444:4444 --name $SELENIUM_DOCKER_NAME -v $PWD:$PWD -v $TMP_DIRECTORY/irida:$TMP_DIRECTORY/irida -v /dev/shm:/dev/shm selenium/standalone-chrome:latest
```
[the remark this answers is not in the document]
Comment (ericenns) [reply run-tests.sh]: Fixed in f8b47d9d61ef9f01ccad61d81867d31d0d49f336

Note: I just set it up to reuse the container if it already exists and wasn't removed from the last run

File: run-tests.sh [hunk 5]
```diff
@@ -109,8 +113,17 @@ test_rest() {
...
+        SELENIUM_OPTS="-Dwebdriver.selenium_url=http://localhost:4444/wd/hub -Djetty.port=33333 -Dserver.base.url=http://$HOSTNAME:33333 -Djava.io.tmpdir=$TMP_DIRECTORY/irida"
+    fi
+	mvn clean verify -B -Pui_testing $SELENIUM_OPTS -Dirida.it.nosandbox=true -Dirida.it.headless=$HEADLESS -Djdbc.url=$JDBC_URL -Dirida.it.rootdirectory=$TMP_DIRECTORY -Dsequence.file.base.directory=$SEQUENCE_FILE_DIR -Dreference.file.base.directory=$REFERENCE_FILE_DIR -Doutput.file.base.directory=$OUTPUT_FILE_DIR -Djdbc.pool.maxWait=$DB_MAX_WAIT_MILLIS $@
 	exit_code=$?
+	if [ "$DO_KILL_DOCKER" = true ]; then docker rm -f -v $SELENIUM_DOCKER_NAME; fi
```
[the remark this answers is not in the document]
Comment (ericenns) [reply run-tests.sh]: Fixed in f8b47d9d61ef9f01ccad61d81867d31d0d49f336

File: run-tests.sh [hunk 6]
```diff
@@ -109,8 +113,17 @@ test_rest() {
...
+    SELENIUM_OPTS=""
+    if [ "$SELENIUM_DOCKER" = false ];
+    then
+        SELENIUM_OPTS="-Dwebdriver.chrome.driver=$CHROME_DRIVER"
+    else
+        docker run -d -p 4444:4444 --name $SELENIUM_DOCKER_NAME -v $PWD:$PWD -v $TMP_DIRECTORY/irida:$TMP_DIRECTORY/irida -v /dev/shm:/dev/shm selenium/standalone-chrome:latest
+        SELENIUM_OPTS="-Dwebdriver.selenium_url=http://localhost:4444/wd/hub -Djetty.port=33333 -Dserver.base.url=http://$HOSTNAME:33333 -Djava.io.tmpdir=$TMP_DIRECTORY/irida"
+    fi
+	mvn clean verify -B -Pui_testing $SELENIUM_OPTS -Dirida.it.nosandbox=true -Dirida.it.headless=$HEADLESS -Djdbc.url=$JDBC_URL -Dirida.it.rootdirectory=$TMP_DIRECTORY -Dsequence.file.base.directory=$SEQUENCE_FILE_DIR -Dreference.file.base.directory=$REFERENCE_FILE_DIR -Doutput.file.base.directory=$OUTPUT_FILE_DIR -Djdbc.pool.maxWait=$DB_MAX_WAIT_MILLIS $@
```
[the remark this answers is not in the document]
Comment (ericenns) [reply run-tests.sh]: Hmm, I never got this can you describe how you caused this error

[the remark this answers is not in the document]
Comment (ericenns) [reply run-tests.sh]: Fixed in 566b1b0d9d3ed2d6a6a9dc638c2f9eaa259581ee

`mkdir -p` will only create the folder if it doesn't exist already

File: run-tests.sh [hunk 7]
```diff
@@ -109,8 +113,17 @@ test_rest() {
...
 
 test_ui() {
-	mvn clean verify -B -Pui_testing -Dwebdriver.chrome.driver=$CHROME_DRIVER -Dirida.it.nosandbox=true -Dirida.it.headless=$HEADLESS -Djdbc.url=$JDBC_URL -Dirida.it.rootdirectory=$TMP_DIRECTORY -Dsequence.file.base.directory=$SEQUENCE_FILE_DIR -Dreference.file.base.directory=$REFERENCE_FILE_DIR -Doutput.file.base.directory=$OUTPUT_FILE_DIR -Djdbc.pool.maxWait=$DB_MAX_WAIT_MILLIS $@
+    SELENIUM_OPTS=""
+    if [ "$SELENIUM_DOCKER" = false ];
+    then
+        SELENIUM_OPTS="-Dwebdriver.chrome.driver=$CHROME_DRIVER"
+    else
+        docker run -d -p 4444:4444 --name $SELENIUM_DOCKER_NAME -v $PWD:$PWD -v $TMP_DIRECTORY/irida:$TMP_DIRECTORY/irida -v /dev/shm:/dev/shm selenium/standalone-chrome:latest
+        SELENIUM_OPTS="-Dwebdriver.selenium_url=http://localhost:4444/wd/hub -Djetty.port=33333 -Dserver.base.url=http://$HOSTNAME:33333 -Djava.io.tmpdir=$TMP_DIRECTORY/irida"
```
[the remark this answers is not in the document]
Comment (ericenns) [reply run-tests.sh]: Fixed in f8b47d9

**Comment by tom114** [comment]
Consistently getting 4 test failures when running with default options `./run-tests.sh --selenium-docker ui_testing`.  Do you see the same?
```
Results :

Tests in error: 
ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT.testAddSamplesToCart(ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT)
  Run 1: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
  Run 2: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
  Run 3: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
  Run 4: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
  Run 5: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
  Run 6: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...

ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT.testSampleSelection(ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT)
  Run 1: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
  Run 2: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
  Run 3: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
  Run 4: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
  Run 5: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
  Run 6: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...

ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSyncPageIT.testSyncProject(ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSyncPageIT)
  Run 1: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
  Run 2: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
  Run 3: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
  Run 4: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
  Run 5: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
  Run 6: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...

ca.corefacility.bioinformatics.irida.ria.integration.remoteapi.CreateRemoteAPIPageIT.testAndConnectToClient(ca.corefacility.bioinformatics.irida.ria.integration.remoteapi.CreateRemoteAPIPageIT)
  Run 1: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
  Run 2: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
  Run 3: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
  Run 4: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
  Run 5: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
  Run 6: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...


Tests run: 175, Failures: 0, Errors: 4, Skipped: 2

```

**Comment by ericenns** [comment]
> Consistently getting 4 test failures when running with default options `./run-tests.sh --selenium-docker ui_testing`. Do you see the same?
> 
> ```
> Results :
> 
> Tests in error: 
> ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT.testAddSamplesToCart(ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT)
>   Run 1: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
>   Run 2: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
>   Run 3: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
>   Run 4: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
>   Run 5: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
>   Run 6: ProjectSamplesPageIT.testAddSamplesToCart:142 » UnsupportedCommand sendKeysToA...
> 
> ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT.testSampleSelection(ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSamplesPageIT)
>   Run 1: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
>   Run 2: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
>   Run 3: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
>   Run 4: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
>   Run 5: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
>   Run 6: ProjectSamplesPageIT.testSampleSelection:127 » UnsupportedCommand sendKeysToAc...
> 
> ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSyncPageIT.testSyncProject(ca.corefacility.bioinformatics.irida.ria.integration.projects.ProjectSyncPageIT)
>   Run 1: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
>   Run 2: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
>   Run 3: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
>   Run 4: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
>   Run 5: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
>   Run 6: ProjectSyncPageIT.setUpTest:26 » Timeout Timed out after 10 seconds waiting fo...
> 
> ca.corefacility.bioinformatics.irida.ria.integration.remoteapi.CreateRemoteAPIPageIT.testAndConnectToClient(ca.corefacility.bioinformatics.irida.ria.integration.remoteapi.CreateRemoteAPIPageIT)
>   Run 1: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
>   Run 2: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
>   Run 3: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
>   Run 4: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
>   Run 5: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
>   Run 6: CreateRemoteAPIPageIT.testAndConnectToClient:70 » Timeout Timed out after 10 s...
> 
> 
> Tests run: 175, Failures: 0, Errors: 4, Skipped: 2
> ```

This issue appears to be caused by the deprecation of the Actions api, from version 3.141.0 and onwards of selenium. https://github.com/SeleniumHQ/selenium/blob/selenium-3.141.0/java/CHANGELOG#L5

I have pinned the version of the docker image to 3.14.0 which passes the tests.

**Comment by ericenns** [comment]
Hi @tom114 Selenium has been upgraded to the latest version and the restriction on the docker container has been lifted. I no longer get test failures when running through docker or using local chromedriver.

**Comment by joshsadam** [comment]
The tests are working now on my Mac, thanks @ericenns

**Pull request merged**